modmail/utils/pagination.py

```diff
@@ -0,0 +1,260 @@
+        self._pages: List[Union[discord.Embed, str]] = []
+        self.source_message = source_message
+        self.prefix = prefix
+        self.suffix = suffix
+        self.max_size = max_size
+        self.linesep = linesep
+        self._embed = embed
+
+        if not isinstance(timeout, (int, float)):
+            raise InvalidArgumentError("timeout must be a float")
+
+        self.timeout = float(timeout)
+        self.clear()
+        for line in contents:
+            self.add_line(line)
+
+        # create the super so the children attributes are set
+        super().__init__()
+
+        # store component states for disabling
+        self.states: Dict[str, Dict[str, Any]] = dict()
+        for child in self.children:
+            attrs = child.to_component_dict()
+            self.states[attrs["custom_id"]] = attrs
+
+    @classmethod
+    async def paginate(
+        cls,
+        contents: Optional[List[str]] = None,
+        source_message: discord.Message = None,
+        /,
+        timeout: float = 180,
+        embed: discord.Embed = None,
+        *,
+        only: Optional[discord.abc.User] = None,
+        channel: discord.abc.Messageable = None,
+        prefix: str = "",
+        suffix: str = "",
+        max_size: int = 4000,
+        linesep: str = "\n",
+        only_users: Optional[List[discord.abc.User]] = None,
+    ) -> None:
+        """Create a paginator, and paginate the provided lines."""
+        paginator = cls(
+            contents,
+            source_message=source_message,
+            timeout=timeout,
+            embed=embed,
+            prefix=prefix,
+            suffix=suffix,
+            max_size=max_size,
+            linesep=linesep,
+            only_users=only_users,
+        )
+
+        if channel is None and source_message is None:
+            raise MissingAttributeError("Both channel and message are None.")
+        elif channel is None:
+            channel = source_message.channel
+
+        paginator.modify_states()
+        if len(paginator.pages) >= 2:
+            msg: discord.Message = await channel.send(
+                content=paginator.pages[paginator._index], view=paginator
+            )
+        else:
+            await channel.send(content=paginator.pages[paginator._index])
+            return
+
+        await paginator.wait()
+        await msg.edit(view=None)
+
+    async def interaction_check(self, interaction: "Interaction") -> bool:
+        """Check if the interaction is by the author of the paginatior."""
+        if self.source_message is None:
+            return True
+        if not (is_valid := self.source_message.author.id == interaction.user.id):
+            await interaction.response.send_message(
+                content="This is not your message to paginate!", ephemeral=True
+            )
+        return is_valid
+
+    def modify_states(self) -> None:
+        """Disable specific components depending on paginator page and length."""
```

Comment:
Could we describe specific components in bit detail here, like if the message is on the first page then back the following buttons would be allowed to have interaction, if it is on the last page then these buttons would be allowed to have interaction.